GIT-49: enable parallel run of test features

harshanarayana opened the pull request:
closes #49 

### TODO

- [x] Add Unit Tests
- [x] Add Documentation
- [x] Add Examples

harshanarayana commented:
@crandles @vladimirvivien Please review this change when time permits and let me know if this is in the right direction.

crandles reviewed (commented):
I think this looks ok but still comparing this against parallel ginkgo / stdlib tests

harshanarayana commented on examples/parallel_features/README.md:
````diff
@@ -0,0 +1,19 @@
+One can also build a test binary and run that with the right arguments.
+
+```bash
+go test -c -o paralles.test .
````
Corrected. Thanks @crandles

harshanarayana commented:
> still comparing this against parallel ginkgo / stdlib tests

I will have to dig into a bit more detail of how `ginkgo` does parallel tests runs inside a given test suite. Been a while since I was in that part of the woods. 

The std benchmarking test however follows a similar approach to what I have here for running parallel tests

```go
for p := 0; p < numProcs; p++ {
		go func() {
			defer wg.Done()
			pb := &PB{
				globalN: &n,
				grain:   grain,
				bN:      uint64(b.N),
			}
			body(pb)
		}()
	}
```
I thought about being able to reuse the `t.Parallel()` helper to get the behavior I wanted, but that did not work as I intended for it to. I am open to suggestion on what else might be a better way to get to achieve this.

vladimirvivien commented on pkg/env/env.go:
```diff
@@ -160,70 +172,134 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
+// AfterEachFeature. This can be invoked with a nillable wg which defines how the test feature is being run.
+// If the wg is set to nil that means the test is being run in sequence and there is no reason to mark a
+// WaitGroup as done
+func (e *testEnv) processTestFeature(wg *sync.WaitGroup, t *testing.T, featureName string, feature types.Feature) {
```
I don't think this should take a workgroup. I suggest to keep generic as possible for clarity (see how in next comment).

vladimirvivien commented on pkg/env/env.go:
```diff
@@ -160,70 +172,134 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
+	}
+
+	var wg sync.WaitGroup
+	for i, feature := range testFeatures {
```
Suggestion: I would separate the execution paths (parallel, non-parallel) cleanly for clarity.  Also, this would make method `e.processTestFeature` truly generic without the need to pass a nil.

```go
if parallelEnabled {
    var wg sync.WaitGroup
    for i, feature := range testFeatures {
        wg.Add(1)
        go func(w *sync.Workgroup){
            defer w.Done()
            e.processTestFeature(t, featName, feature)
          }(&wg)
    }
    wg.Wait()
}else{
    for i, feature := range testFeatures {
         e.processTestFeature(t, featName, feature)
    }
}
```

harshanarayana replied:
You have a valid point. The `porcessTestFeature` can/should indeed be agnostic of the `WaitGroup` and the sequence/parallel workflow. What you suggested here looks better. Let me update it accordingly.

harshanarayana replied:
```golang
for i, feature := range testFeatures {
		featName := feature.Name()
		if featName == "" {
			featName = fmt.Sprintf("Feature-%d", i+1)
		}
		if runInParallel {
			wg.Add(1)
			go func(wg *sync.WaitGroup) {
				defer wg.Done()
				e.processTestFeature(t, featName, feature)
			}(&wg)
		} else {
			e.processTestFeature(t, featName, feature)
		}
	}
	if runInParallel {
		wg.Wait()
	}
```

turned it like above to avoid duplication of loop + the feature name generation bit.

vladimirvivien commented on pkg/env/env.go:
```diff
@@ -160,70 +172,134 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
+	e.processTestActions(t, afterTestActions)
+}
+
+// TestInParallel executes a series a feature tests from within a
```
❤️ the comments on these methods.

vladimirvivien reviewed (commented):
This is an awesome start and thank you for the effort.
I left some comments.

harshanarayana commented on pkg/env/env.go:
```diff
@@ -160,70 +172,134 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
+
+	e.processTestActions(t, beforeTestActions)
+
+	if e.cfg.ParallelTestEnabled() && enableParallelRun {
```
Sure. I will take care of it.

harshanarayana commented:
> This is an awesome start and thank you for the effort. I left some comments.

I have addressed the review comments you have. Please check then when time permits. Let me know if you see any other changes required as well

ShwethaKumbla commented:
/ok-to-test

vladimirvivien commented on pkg/env/env.go:
```diff
@@ -160,70 +172,133 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
-// BeforeTest and AfterTest operations are executed before and after
-// the feature is tested respectively.
-func (e *testEnv) Test(t *testing.T, testFeatures ...types.Feature) {
+// panicOnMissingContext is used to check if the test Env has the right context setup
```
Nit: it checks not for the `right` context, but that there is a `non-nil` context.

vladimirvivien reviewed (commented):
@harshanarayana This is great work and progress on this PR.
Couple of things:
* Add test in pkg/envconf/config_test.go
* Add test in pkg/flags/flags_test.go

Also left comment.

harshanarayana commented:
@vladimirvivien  Happy new year. 

I just got back to working mode and will take a look at this today and get your comments addressed.

harshanarayana commented:
@vladimirvivien  I've updated the PR and added tests for the two files that were missing earlier in the PR.

vladimirvivien reviewed (approved):
Thanks @harshanarayana . This looks awesome.
